src/klein/_headers.py

```diff
@@ -0,0 +1,315 @@
+    of C{(name, value)} L{tuple}s.
+
+    A dictionary-like interface that maps text names to an ordered sequence of
+    text values.
+    This interface assumes that both header name bytes and header value bytes
+    are encoded as ISO-8859-1.
+
+    Note that header name bytes should be strictly encoded as ASCII; this
+    interface uses ISO-8859-1 to provide interoperability with (naughty) HTTP
+    implementations that send non-ASCII data.
+    Because ISO-8859-1 is a superset of ASCII, this will still work for
+    well-behaved implementations.
+    """
+
+    rawHeaders = Attribute(
+        """
+        Raw header data as a tuple in the from: C{((name, value), ...)}.
+        C{name} and C{value} are bytes.
+        Headers are provided in the order that they were received.
+        Headers with multiple values are provided as separate name and value
+        pairs.
+        """
+    )  # type: RawHeaders
+
+
+    def getValues(name):
+        # type: (AnyStr) -> Iterable[AnyStr]
+        """
+        Get the values associated with the given header name.
+
+        If the given name is L{bytes}, the value will be returned as the raw
+        header L{bytes}.
+
+        If the given name is L{Text}, the name will be encoded as ISO-8859-1
+        and the value will be returned as text, by decoding the raw header
+        value bytes with ISO-8859-1.
+
+        @param name: The name of the header to look for.
+
+        @return: The values of the header with the given name.
+        """
+
+
+
+class IMutableHTTPHeaders(IHTTPHeaders):
+    """
+    Mutable HTTP entity headers.
+    """
+
+    def remove(name):
+        # type: (AnyStr) -> None
+        """
+        Remove all header name/value pairs for the given header name.
+
+        If the given name is L{Text}, it will be encoded as ISO-8859-1 before
+        comparing to the (L{bytes}) header names.
+
+        @param name: The name of the header to remove.
+        """
+
+
+    def addValue(name, value):
+        # type: (AnyStr, AnyStr) -> None
+        """
+        Add the given header name/value pair.
+
+        If the given name is L{bytes}, the value must also be L{bytes}.
+
+        If the given name is L{Text}, it will be encoded as ISO-8859-1, and the
+        value, which must also be L{Text}, will be encoded as ISO-8859-1.
+        """
+
+
+
+# Encoding/decoding header data
+
+HEADER_NAME_ENCODING  = "iso-8859-1"
+HEADER_VALUE_ENCODING = "iso-8859-1"
+
+
+def headerNameAsBytes(name):
+    # type: (String) -> bytes
+    """
+    Convert a header name to bytes if necessary.
+    """
+    if isinstance(name, bytes):
+        return name
+    else:
+        return name.encode(HEADER_NAME_ENCODING)
+
+
+def headerNameAsText(name):
+    # type: (String) -> Text
+    """
+    Convert a header name to text if necessary.
+    """
+    if isinstance(name, Text):
+        return name
+    else:
+        return name.decode(HEADER_NAME_ENCODING)
+
+
+def headerValueAsBytes(value):
```

Comment:
Do we really need this pretense that we the name and the value can be different bad encodings? :)

Reply:
It is definitely the case, based on specs, that the encodings can be different.

This doesn't currently support alternative header value encodings, because the spec is kind of confusing about it, but I didn't want to ignore the possibility of doing that in the future, so I kept them separate.